Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163652

This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, namely  Qwen/Qwen2-1.5B-Instruct it makes reasoning logic easier to do those replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

laithsakka committed:
allow unbacked to unbacked replacements if rhs unbacked are all inputs

[ghstack-poisoned]

laithsakka committed:
Update on "allow unbacked to unbacked replacements if rhs unbacked are all inputs"




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "allow unbacked to unbacked replacements if rhs unbacked are all inputs"




cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. But i think the this PR is still valid, we do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. But i think the this PR is still valid, we do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

cc ezyang EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. But i think the this PR is still valid, we do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. But i think the this PR is still valid, we do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

ezyang commented:
I haven't reviewed the code but the high level reasoning LGTM

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

aorenste reviewed (commented):
Code looks reasonable but some tests are failing

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -4913,8 +4924,18 @@ def create_unspecified_symbol(
+
     @record_shapeenv_event()
-    def create_symbol(
+    def create_non_data_dependent_symbol(
```
mm i mean there is no much places were we distinguish those

so we have
backed : non_data_dependent
unbacked
  - inputs: non_data_dependent
  - non inputs: data dependent. (item, tolist ..etc)
  
for the  purpose of creating symbols this function was named create_symbol, which is very general. This is really meant for non_data_dependent symbols.

bobrenjc93 replied:
This seems so overly complicated. Can you explain why we can't just have create_backed_symbol and create_unbacked_symbol where unbacked are implicitly data dependent?

If the only concern is around the input striding, I'd much prefer banning on contiguous striding on unbacked vs all of this additional complexity around backed symbols, unbacked symbols, data dependent symbols, non data depdenent symbols...

laithsakka replied:
we already have this API called
create_symbol now, i do not want people to use it to create non input symbols it will fail
does that make sense?
just renamed it to create_non_data_dependent_symbol , is create_input_symbol better?

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, it makes reasoning logic easier to do those
replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]

laithsakka committed:
Update on "Allow unbacked to unbacked replacements if rhs unbacked symbols are all inputs"


This partially solve the issue https://github.com/pytorch/pytorch/issues/163641. We do not need to ban unbacked to unbacked replacement if all rhs symbols are inputs since we know those symbols are seen by the whole program.

This issue was found as i was tracing some vllm models with unbacked, namely  Qwen/Qwen2-1.5B-Instruct it makes reasoning logic easier to do those replacements. 

as for data dependent similar pattern, I am thinking to create a set of replacements that we apply only during static eval
instead of none. to make reasoning better. 

I am also renaming create_symbol  to create_non_data_dependent_symbol, this function now(even before this PR) is only used to create 
no data dependent symbols and should not be used for the purpose of creating data dependent symbols. 
so be explicit is better. 
I kept  create_symbol for backward compatibility and added deprecation annotation so we can remove it. 
cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben Lucaskabela

[ghstack-poisoned]
